Resolution.fromWeb3Provider

JohnnyJumper opened the pull request:
continuation of  #67.


this should be better than the #67 attempt. I also added tests checking following providers: 
[web3-providers-http](https://github.com/ethereum/web3.js/tree/242bfd0adfb8581efd79ef6589d782fbc1b466d8/packages/web3-providers-http)
[web3-providers-ws](https://github.com/ethereum/web3.js/tree/242bfd0adfb8581efd79ef6589d782fbc1b466d8/packages/web3-providers-ws)

I wasn't able to connect an ipc provider from [here](https://github.com/ethereum/web3.js/tree/242bfd0adfb8581efd79ef6589d782fbc1b466d8/packages/web3-providers-ipc)


Also, I want to deprecate the web3Provider property since it is not actually a web3provider, but not sure how to start this. Any thoughts?

mvlabat commented on src/Resolution.ts:
```diff
@@ -108,6 +110,33 @@ export default class Resolution {
+   * Creates a resolution instance from AbstractProvider
+   * @param provider - callback base provider
+   */
+  static fromWeb3Provider(provider: AbstractProvider): Resolution {
```
I guess we also can introduce `fromEthersProvider` method for convenience

mvlabat commented on src/Resolution.ts:
```diff
@@ -108,6 +110,33 @@ export default class Resolution {
+  static fromWeb3Provider(provider: AbstractProvider): Resolution {
+    const customProvider: Web3Provider = {
+      sendAsync: (method, params) => new Promise((resolve, reject) => {
+        if (provider.sendAsync != undefined) {
```
```suggestion
        if (provider.sendAsync !== undefined) {
```
(and below as well)

mvlabat commented on src/resolutionError.ts:
```diff
@@ -49,6 +50,7 @@ const HandlersByCode = {
     recordName: string;
     domain: string;
   }) => `No ${params.recordName} record found for ${params.domain}`,
+  [ResolutionErrorCode.IncorectProvider]: (params: {}) => "Provider doesn't have sendAsync or send function."
```
```suggestion
  [ResolutionErrorCode.IncorectProvider]: (params: {}) => "Provider doesn't implement sendAsync or send method"
```
Not forcing this change, but I thought this might be a better wording

mvlabat commented on src/utils/testHelpers.ts:
```diff
@@ -82,9 +82,8 @@ export function mockAPICalls(testName: string, url = MainnetUrl) {
  * UNSTOPPABLE_RESOLUTION_INFURA_PROJECTID env variable if any
  */
-export function secretInfuraLink(): string {
+export function secretInfuraLink(options: {wss: boolean} = {wss: false}): string {
```
I think we'd better use enum here as well, maybe in future we'll see some other protocols emerging

JohnnyJumper commented on src/utils/testHelpers.ts:
```diff
@@ -82,9 +82,21 @@ export function mockAPICalls(testName: string, url = MainnetUrl) {
-  if (secret) url = `https://mainnet.infura.io/v3/${secret}`;
-  return url;
+  let url = "";
+  switch (options) {
```
Is there any better-looking way to work with enums beside the switch case? This looks too big in my opinion for such a small change.

bogdan replied:
Make a map like:

```
const url = {wss: 'url1', http: 'url2'}[options]
```

JohnnyJumper replied:
So simple and elegant! Added this way. Strange I haven't thought of that before! Thanks!

bogdan commented on src/types.ts:
```diff
@@ -104,17 +104,37 @@ export type Blockchain =
+
+export interface AbstractProvider {
+  sendAsync?: (payload: JsonRpcPayload, callback: (error: Error | null, result?: JsonRpcResponse) => void) => void;
+  send?: (payload: JsonRpcPayload, callback: (error: Error | null, result?: JsonRpcResponse) => void) => void;
```
Woun't there be one required property here or maybe there should be at least one? You can describe this situation with typescript like:

```
type ProviderMethod = (payload: JsonRpcPayload, callback: (error: Error | null, result?: JsonRpcResponse) => void) => void
type AbstratProvider = {sendAsync: ProviderMethod} | {send: ProviderMethod}
```

JohnnyJumper replied:
It seems that the newer providers have been migrated from using send method to sendAsync method.
That means depending on which provider the client is using they might either not have sendAsync at all and have send method or vise-versa 

If I do it in a way above the user will be getting typescript error saying that the provider I am sending is not the same as being expected since sendAsync is not implemented for older providers like  web3-providers-ws or web3-providers-http  and it won't compile for them without hard type casting.

I think the method we used now is the safest since we do manually check for both methods anyway and in case none found we throw the IncorectProvider Error with the message that neither send nor sendAsync is implemented in the provider.

JohnnyJumper replied:
To be more clear this is what I am getting on our side when defining AbstractProvider via | sign, I would assume the client-side would get a similar error when sending a provider to us

<img width="1012" alt="Screen Shot 2020-06-11 at 4 04 12 PM" src="https://user-images.githubusercontent.com/16522024/84447615-50dda500-abfd-11ea-82f3-571e6bca26fa.png">

Or maybe I am not doing something right.

mvlabat replied:
@bogdan Some of the deprecated web3 providers (like `web3-providers-http` that @JohnnyJumper mentioned, not `web3-providers-http` though) implement only `send` method. See https://github.com/ethereum/web3.js/blob/1.x/packages/web3-core-requestmanager/src/index.js#L162

mvlabat commented on src/utils/testHelpers.ts:
```diff
@@ -82,9 +82,16 @@ export function mockAPICalls(testName: string, url = MainnetUrl) {
  * UNSTOPPABLE_RESOLUTION_INFURA_PROJECTID env variable if any
  */
-export function secretInfuraLink(): string {
+export function secretInfuraLink(option: InfuraProtocolOptions = InfuraProtocolOptions.http): string {
```
`option` - let's make this name more self-descriptive, like `infuraProtocol`.
Also I don't fancy `Options` prefix in the name of the enum type, it seems to be a bit irrelevant for enums

mvlabat commented on src/resolutionError.ts:
```diff
@@ -18,6 +18,7 @@ export enum ResolutionErrorCode {
   UnsupportedCurrency = 'UnsupportedCurrency',
   IncorrectResolverInterface = 'IncorrectResolverInterface',
   RecordNotFound = 'RecordNotFound',
+  IncorectProvider = 'IncorectProvider'
```
Missing 'r' in the word "Incorect"

mvlabat commented on src/types.ts:
```diff
@@ -104,17 +104,37 @@ export type Blockchain =
 };
 
-export interface Web3Provider {
+export interface Provider {
```
This is a breaking change. If you want to rename it, export an alias `Web3Provider` as well

JohnnyJumper replied:
I added an export but never used it anywhere. I guess i can just mark it as deprecated to remove in future versions?

mvlabat commented on src/Resolution.test.ts:
```diff
@@ -1,11 +1,14 @@
 import nock from 'nock';
 import Resolution, { ResolutionErrorCode } from './index';
 import { UnclaimedDomainResponse, NamingServiceName } from './types';
+var Web3WsProvider = require('web3-providers-ws');
+var Web3HttpProvider = require('web3-providers-http');
```
Why `var` and `require` though?

JohnnyJumper replied:
I tried to make it with an import instead of require and typescript couldn't understand the Web3HttpProvider, I tried to import HttpProvider as in class name and got errors that it doesn't have a constructor =-( So I reverted to require

mvlabat replied:
Maybe `import Web3HttpProvider = require('web3-providers-http');` will work?

mvlabat replied:
Seems like they export them in CommonJS way, so it should work

JohnnyJumper replied:
This is what i got whenever i change to import. 
I tried with 
`Maybe import Web3HttpProvider = require('web3-providers-http');`
Same issue
<img width="1144" alt="Screen Shot 2020-06-15 at 11 57 46 AM" src="https://user-images.githubusercontent.com/16522024/84695280-71f50d00-aeff-11ea-95a9-6fe633c0067d.png">

mvlabat commented on src/ens.ts:
```diff
@@ -37,8 +37,8 @@ export default class Ens extends EthereumNamingService {
    */
-  constructor(source: NamingServiceSource = true, web3Provider?: Web3Provider) {
-    super(web3Provider);
+  constructor(source: NamingServiceSource = true, provider?: Provider) {
```
The constructor misses documentation about `provider` argument.
Also we don't validate `provider` here, though we validate it in the `fromWeb3Provider` method. The behaviour should be consistent

mvlabat replied:
I guess this constructor assumes passing etherjs provider only, we should explicitly mention it in the documentation, cause it can be confusing for the users

JohnnyJumper replied:
Ens.ts is for internal use only, and Resolution has all the constructors that deal with the conversion of providers to make sure it is consistent. I added the docs to it as well

mvlabat commented on src/types.ts:
```diff
@@ -104,17 +104,44 @@ export type Blockchain =
+/**
+ * @deprecated
+ */
+export interface Web3Provider {
+  sendAsync: (method: string, params: any) => Promise<any>; 
+}
```
```suggestion
export type Web3Provider = Provider;
```
to eliminate code duplication

mvlabat commented on src/resolutionError.ts:
```diff
@@ -18,6 +18,7 @@ export enum ResolutionErrorCode {
   UnsupportedCurrency = 'UnsupportedCurrency',
   IncorrectResolverInterface = 'IncorrectResolverInterface',
   RecordNotFound = 'RecordNotFound',
+  IncorrectProvider = 'IncorrectProvider'
```
nitpick: missing comma

mvlabat commented on src/utils/contract.ts:
```diff
@@ -99,9 +99,9 @@ export default class Contract extends BaseConnection {
+    if (this.provider) {
+      return await this.provider
         .sendAsync('eth_call', params)
         .then(resp => resp.json());
```
Web3 provider responses have different API (they don't have it even, as they come as parsed jsons). So this doesn't work unfortunately

JohnnyJumper replied:
I have added an if there in case if the provider is custom as the one we implemented in previous tests 
```const provider = {
        sendAsync: (method: string, params: any) => {
          return nodeFetch(secretInfuraLink(), {
            method: 'POST',
            body: JSON.stringify({
              method,
              params,
              jsonrpc: '2.0',
              id: 1,
            }),
          });
        },
      };```

and after removing all the mocks internally it passes everything again.

mvlabat commented on src/Resolution.test.ts:
```diff
@@ -359,5 +362,30 @@ describe('Resolution', () => {
+    it('should work with web3HttpProvider', async () => {
+      const provider = new Web3HttpProvider(secretInfuraLink());
+      const resolution = Resolution.fromWeb3Provider(provider);
+      const eyes = mockAsyncMethods(resolution.cns, {
+        getResolver: '0xb66DcE2DA6afAAa98F2013446dBCB0f4B0ab2842',
+        getRecord: '0x8aaD44321A86b170879d7A244c1e8d360c99DdA8'
+      });
```
These tests won't be able to test anything if we mock CNS functions directly - the provider flow isn't covered

JohnnyJumper replied:
Wow that is so true. I start getting the issues with web3 providers after I remove these mocks.

JohnnyJumper commented on src/utils/contract.ts:
```diff
@@ -102,8 +102,8 @@ export default class Contract extends BaseConnection {
-        .sendAsync('eth_call', params)
-        .then(resp => resp.json());
+      .sendAsync('eth_call', params)
+      .then(resp => resp.json !== undefined ? resp.json() : resp);
```
I am not sure if this is the best way to solve the issue but after this change, the test are being passed.

bogdan replied:
You need to solve this when you make a wrapper inside `Resolution.fromWeb3Provider`.

bogdan commented on src/resolutionError.ts:
```diff
@@ -18,6 +18,7 @@ export enum ResolutionErrorCode {
   UnsupportedCurrency = 'UnsupportedCurrency',
   IncorrectResolverInterface = 'IncorrectResolverInterface',
   RecordNotFound = 'RecordNotFound',
+  IncorrectProvider = 'IncorrectProvider',
```
It is better to avoid mixing resolution exceptions into the configuration excetions. You may create a `ConfigurationError` for this or just avoid a custom class for now. It is good idea to just separate expctions occur when library is used from exceptions occur when library is configured.

bogdan commented on src/types.ts:
```diff
@@ -104,15 +104,48 @@ export type Blockchain =
+}
+
+export interface OldAbstractProvider {
+  send: (payload: JsonRpcPayload, callback: (error: Error | null, result?: JsonRpcResponse) => void) => void;
```
```suggestion
  send: ProviderMethod,
```

mvlabat commented:
Ahem, I've just discovered a class in ethers.js library... It's called `Web3Provider`, and here's its source code:
https://github.com/ethers-io/ethers.js/blob/1a4f7d1b53050b69a85b1afbab11a2761f22f5bb/packages/providers/src.ts/web3-provider.ts#L77

This is how it can be used to wrap web3 providers: `new ethers.providers.Web3Provider(provider)`

I'm now asking myself whether we needed this feature at all 😶

mvlabat commented on src/utils/contract.ts:
```diff
@@ -103,7 +103,7 @@ export default class Contract extends BaseConnection {
       return await this.provider
       .sendAsync('eth_call', params)
-      .then(resp => resp.json !== undefined ? resp.json() : resp);
+      .then(resp => resp.json());
```
`JsonRpcProvider::send` responses don't have `json` method either. Seems like this code never worked even for ethers.js

JohnnyJumper commented:
According to the newest ethers [docs](https://docs.ethers.io/v5/api/providers/provider/) their provider doesn't give me a method to talk to a smart contract freely. Instead, it seems they are using [jsonRpcProvider](https://docs.ethers.io/v5/api/providers/jsonrpc-provider/)

I have fixed fromJsonRpcProvider and tested it with one from ethers library

Also, I have removed fromEthersProvider. Seems it didn't even work and I couldn't found any proof that somebody was using this method, so decided to remove here while there aren't any users. (P.S should I update the changelog accordingly or return the fromEthersProvider and if so which exactly provider should it wait for?)

There is a huge discussion about (EIP-1193)[https://eips.ethereum.org/EIPS/eip-1193] over [here](https://github.com/ethereum/EIPs/issues/2319) The last comment was made 7 days ago and there are lot of projects that implemented it already including metamask and ethers. Here is the [PR](https://github.com/ethereum/EIPs/pull/2699) 

The list of supported projects:
So far, the following projects support or have implemented this EIP:

- ethers.io (@ricmoo)
- eth-provider (@floating)
- buidler.dev (@alcuadrado)
- MetaMask (@rekmarks)
- web3.js (@ryanio @frozeman)
- WalletConnect (@pedrouid)
- rainbow.me (@brunobar79)

And looks like this is going to be a standard sooner than later, but I might be mistaken. In that case, I introduced  fromEIP1193Provider (any better names?) which works similarly as per [ethers-docs](https://docs.ethers.io/v5/api/providers/other/#Web3Provider--ExternalProvider) It looks for .request method which is the new standard and in case it couldn't go for send or sendAsync depends which one is being found. 

P.s. I am pretty sure there are going to be some suggestions on how to make it better. 
and as a question, I am not really sure how to test the EIP1193 providers:? 

@bogdan  I tried to mock the send from the provider just as you showed me, but that didn't really work. I looked into [web3-http-provider send method](https://github.com/NicolasMahe/web3-providers-http/blob/874aba3ef134d2d2394e05738cdae432c0496a76/src/index.js#L107) and tried to mock that but got lost a little. This function doesn't really return anything O_o Maybe I missed something?

rekmarks commented:
:wave:

I’m one of the authors of 1193. All outstanding issues with 1193 have been resolved at this time, and our review period ended 3 weeks ago.

We’re merely waiting for a final review from an EIP editor. It’s not final until it’s final, but we expect any feedback to be purely procedural.

Speaking for MetaMask, we’re shipping 1193 into production any day now, with our existing provider API intact.

mvlabat commented on src/types.ts:
```diff
@@ -86,15 +86,66 @@ export type Blockchain =
+}
+
+export interface Provider {
+  call: (method: string, params: ProviderParams) => Promise<any>;
```
I'm thinking now that maybe it's worth to just introduce EIP-1193 interface to our library and wrap all the supported adapters into it, instead of introducing "YetAnotherProviderInterface"?

JohnnyJumper replied:
It sounds like a good idea! When EIP-1193 will go out we will already be supporting that out of the box

mvlabat commented on src/types.ts:
```diff
@@ -86,15 +86,66 @@ export type Blockchain =
+  callback: (error: Error | null, result?: JsonRpcResponse) => void,
+) => void;
+
+export interface OldAbstractProvider {
```
I'd rename it to something that gives us some context that this is web3 interface. `Web3Version0Provider` maybe or something like that?

mvlabat commented on src/types.ts:
```diff
@@ -86,15 +86,66 @@ export type Blockchain =
 
-export interface Web3Provider {
-  sendAsync: (method: string, params: any) => Promise<any>;
+export interface ExternalProvider {
```
Shall we rename it to `EIP1193Provider` too?

mvlabat commented on src/zns.ts:
```diff
@@ -57,10 +60,14 @@ export default class Zns extends NamingService {
     this.url = source.url as string;
     if (!this.network) {
-      throw new Error('Unspecified network in Resolution ZNS configuration');
+      throw new ConfigurationError(ConfigurationErrorCode.UnspecifiedNetwork, {
```
Do we consider this a breaking change?

JohnnyJumper replied:
Shouldn't be, before it was a simple error now it is a Configuration Error which extends Error. These errors were designed to be raised to tell a developer that something is wrong before they could launch. Meaning they were not supposed to be caught

JohnnyJumper commented on src/Resolution.ts:
```diff
@@ -96,16 +104,117 @@ export default class Resolution {
+   * This should create an instance of resolution with any EIP-1193 compatable provider.
+   * See this: https://eips.ethereum.org/EIPS/eip-1193
+   * In case if provider doesn't implement request method,
+   * we check for send(payload, callback) method and sendAsync(payload, callback)
```
I followed the ethers library from here https://docs.ethers.io/v5/api/providers/other/#Web3Provider--ExternalProvider
But now I see why this is redundant. We do have web3VersionXProvider for such things. Should I throw a configuration error that suggest a different method based on which method is defined in the provider?

mvlabat commented on src/Resolution.ts:
```diff
@@ -94,18 +101,97 @@ export default class Resolution {
+  /**
+   * Create a resolution instance from web3 0.x version provider
+   * @param provider - an 0.x version provider from web3 ( must implement sendAsync(payload, callback) )
+   * see https://github.com/ethereum/web3.js/blob/0.20.7/lib/web3/httpprovider.js#L116
```
```suggestion
   * @see https://github.com/ethereum/web3.js/blob/0.20.7/lib/web3/httpprovider.js#L116
```

mvlabat commented on src/Resolution.ts:
```diff
@@ -94,18 +101,97 @@ export default class Resolution {
   /**
    * Creates a resolution instance with configured provider
-   * @param provider - any provider with sendAsync function impelmented
+   * @param provider - any provider compatable with EIP-1193 (https://eips.ethereum.org/EIPS/eip-1193)
```
```suggestion
   * @param provider - any provider compatible with EIP-1193 (https://eips.ethereum.org/EIPS/eip-1193)
```

mvlabat commented on src/types.ts:
```diff
@@ -86,15 +86,67 @@ export type Blockchain =
+  params?: ProviderParams;
+}
+
+export interface Provider {
+  request: (request: RequestArguments) => Promise<any>;
+}
```
We now have to equal providers: `Provider` and `EIP1193Provider`

mvlabat commented on src/types.ts:
```diff
@@ -86,15 +86,67 @@ export type Blockchain =
+  request: (request: RequestArguments) => Promise<any>;
+}
+
+/**
+ * @deprecated
+ */
+export type Web3Provider = Provider;
```
This isn't a correct alias any longer. If we are breaking the old functionality (that wasn't working anyway), probably it's just better to get rid of it. What do you think @bogdan?

mvlabat commented on src/Resolution.ts:
```diff
@@ -94,18 +101,97 @@ export default class Resolution {
    */
-  static provider(provider: Web3Provider): Resolution {
-    return new this({ blockchain: { web3Provider: provider } });
+  static provider(provider: Provider): Resolution {
```
If this method has got its interface changed, I'm in for renaming it to `fromProvider` to make it more consistent with others

mvlabat commented on src/Resolution.ts:
```diff
@@ -94,18 +101,97 @@ export default class Resolution {
    * Creates a resolution instance from configured jsonRPCProvider
    * @param provider - any jsonRPCprovider will work as long as it's prototype has send(method, params): Promise<any> method
    */
   static jsonRPCprovider(provider): Resolution {
```
I guess it was never working as designed well, so I'd also take opportunity to rename it to `fromEthersJsonRpcProvider`

JohnnyJumper commented:
Not ready yet. Need to find a better way to test for arguments in tests and check the errors working appropriately

mvlabat commented on src/utils/oldProvider/oldHttpProvider.ts:
```diff
@@ -0,0 +1,157 @@
+/*
```
I don't think we should copy-paste web3.js code into our library. If it's used only for testing, why don't we add web3 as a dev dependency?

mvlabat replied:
@bogdan we can bring several versions of the same dependency using package aliases
https://stackoverflow.com/a/56495651/4284929
I haven't tested it myself, but I believe it should work (we can try at least)

0.x being not split into subpackages shouldn't be a problem, as long as it's a dev dependency

bogdan replied:
Well, lets try that if it could be easy enough. I don't believe in the flexibility of package.json because disappointed me a lot of times. In this particular case,  there might problems with a dependencies of the dependencies.

JohnnyJumper replied:
It worked!

mvlabat commented on src/utils/testHelpers.ts:
```diff
@@ -3,7 +3,9 @@ import _ from 'lodash';
 import mockData from '../testData/mockData.json';
+import {default as HttpProvider} from './oldProvider/oldHttpProvider';
 
+export {HttpProvider as oldWeb3Provider};
```
1.x may become "old" too in future. We should be more specific, even if it's only our internal code. I suggest we stick to our already existing convention: "Web3Version0"

mvlabat commented on CHANGELOG.md:
```diff
@@ -1,4 +1,6 @@
 ## Unreleased
+* Added web3 providers supports such as http and ws protocols
```
```suggestion
* Added support of web3.js and ethers.js providers
```

mvlabat commented on CHANGELOG.md:
```diff
@@ -1,4 +1,6 @@
 ## Unreleased
+* Added web3 providers supports such as http and ws protocols
+* Introduced ConfigurationError which handles all error that were made cause of falty configurations
```
I believe that errors can't handle other errors (lexical mistake)

```suggestion
* Introduced ConfigurationError for validating method arguments
```
Still dunno about this exact wording, might sound a bit too abstract

bogdan replied:
You need to describe why it is important to the user, like:

```suggestion
* Throw `ConfigurationError` instead of basic `Error` when Resolution library is configured incorrectly
```

In this way I know what I can do with it: I can target it specifically within `catch` block. That should be obvious, but JS developers are not that profound in error handling so we can even suggest that in a changelog.

mvlabat commented on src/Resolution.ts:
```diff
@@ -44,68 +54,127 @@ export default class Resolution {
    */
-  static provider(provider: Web3Provider): Resolution {
-    return new this({ blockchain: { web3Provider: provider } });
+  static fromEipEthProvider(provider: Provider): Resolution {
```
```suggestion
  static fromEip1193Provider(provider: Provider): Resolution {
```
Currently it stands for `from Ethereum Improvement Proposal Eth Provider`, which imo is tautology and isn't specific enough (what if there appears another proposal for web3 interface?)

mvlabat commented on package.json:
```diff
@@ -62,6 +63,9 @@
     "ts-node": "^8.6.2",
     "typedoc": "0.15.4",
     "typescript": "3.7.3",
+    "web3-0.2.7": "npm:web3@0.2.7",
```
It's 0.20.7

mvlabat commented on src/Resolution.ts:
```diff
@@ -44,68 +55,116 @@ export default class Resolution {
+  static fromEthersProvider(provider: EthersProvider): Resolution {
+    if (provider.call === undefined) throw new ConfigurationError(ConfigurationErrorCode.IncorrectProvider);
+    return this.fromEip1193Provider({
+      request: async (request: RequestArguments) => await provider.call(request.params![0])
```
I think it's worth to add a guard here checking a request method. Since this the code always executes `call` method, we should at least check that `request.method` equals `eth_call`, so we don't accidentally try to call another method and receive an unexpected API error.
```
  if (request.method !== 'eth_call') {
    throw new UnsupportedMethod() // something like that
  }
```

Delegat43 commented:
5403257
